### PR TITLE
Deal more robustly with server responses

### LINKED ISSUE #110: Accept either hashes or IDs for foreign key fields
Entities may be related to each other via one to one or one to many relationships. For example, a product has a one to one relationship with an organization and a one to many relationship with repositories. In other words, a product may belong to one organization and may have many repositories.

The server typically describes these relationships using hashes of data. For example, if one reads a lifecycle environment by issuing an HTTP GET request to `/katello/api/v2/environments/:id`, the server will return a JSON payload in this form:

```json
{
    "name": "…",
    "description": "…",
    "organization": {"id": …, "name": "…"},
}
```

Similarly, one to many relationships are typically described like so:

```json
{
    "organizations": [
        {"id": …, "name": "…"},
        {"id": …, "name": "…"},
    ]
}
```

This is the typical way that relationships are described. But sometimes the server returns IDs instead. For example, if you issue an HTTP GET request to `/katello/api/v2/products/:id`, the server will return a JSON payload in this form:

```json
{
    "gpg_key_id": …,  # one to one
    "sync_plan_id": …,
    "organization": {"id": …, "name": "…"},
}
```

Notice how the one to one relationships on a product are represented using a mixture of IDs and hashes. There's no apparent rhyme or reason here. And one to many relationships may be represented like so:

```json
{
    "related_entity_ids": […, …, …],
}
```

NailGun currently deals with this weirdness by assuming that all relationships are described by hashes and adding in hacks as necessary whenever anything else is done. This makes NailGun less robust and is the cause of much repetitive code in module `nailgun.entities`. Rather than hacking around the hash/ID discrepancies on an individual basis, NailGun should search for both IDs and hashes when trying to populate foreign key fields on entities. Doing so will let NailGun be more robust and let us remove much redundant code from module `nailgun.entities`.

### LINKED ISSUE #114: Accept pluralized fields from the server
Entities may have one to many relationships with one an other, and the names used to describe these fields can vary. For example, the `Organization` entity has a one to many relationship with the `Media` and `Subnet` entities via the "media" and "subnet" fields. An HTTP GET request to `/katello/api/v2/organizations/:id` returns a JSON payload like this:

```json
{
    "media": [{"id": 1}, {"id": 2}],
    "subnets": [{"id": 3}, {"id": 4}],
    …
}
```

Similarly, the `Location` entity has a one to many relationship with the `SmartProxy` entity via the "smart_proxy" field. An HTTP GET request to `/api/v2/locations/:id` returns a JSON payload like this:

```json
{
    "smart_proxies": [{"id": 3}, {"id": 4}],
    …
}
```

In the above example, there are two different spellings for a field:

* The field-name as-is.
* A pluralized version of the field name.

And as described in #110, a one to many relationship may be described with a name in the form `<field_name>_ids`.

NailGun should be capable of dealing with the pluralized version of a field name.

### PR DESCRIPTION
**Pull request by Ichimonji10**
Fix #110:

> Rather than hacking around the hash/ID discrepancies on an individual basis,
> NailGun should search for both IDs and hashes when trying to populate foreign
> key fields on entities. Doing so will let NailGun be more robust and let us
> remove much redundant code from module `nailgun.entities`.

Fix #114:

> NailGun should be capable of dealing with the pluralized version of a field
> name.

Test results against downstream:

    $ make test-foreman-api
    python -m cProfile -o test-foreman-api.pstats $(which nosetests) --logging-filter=nailgun,robottelo --with-xunit --xunit-file=foreman-results.xml tests/foreman/api
    ......SSS.....................S....................E......................SSSSSSSSSSSSSSSSSSSSSSSS........SSSSSSSSSSSSSSSSSSSSSSSSS..............SSSSSSSSSSSSS.............S.SSSSSSS.SSSSSS................................................S.........SSSSSSSSSSSSS...S..S...S..SSSSSSSSSS.SSSSSSSS.....S...S......S.....................................................................S...S.......................................................................................S.............................................................SSSSSS..S.SS.SSSSS.S..SSS.SSSSSSSSSSSSSSSSSSSSSSSSSSSSS.S..................S..S...........................................................SE.........................................................SSSSSSSS...SSSSSSSSSSSSSSSSSSSSSSSSSSSSSS..S.SSS.SSSSSSSSSSSSSSSSSSSSSS............................................
    ======================================================================
    ERROR: test suite for <class 'tests.foreman.api.test_contentview.CVRedHatContent'>
    ----------------------------------------------------------------------
    Traceback (most recent call last):
    File "/home/ichimonji10/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/suite.py", line 209, in run
        self.setUp()
    File "/home/ichimonji10/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
        self.setupContext(ancestor)
    File "/home/ichimonji10/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
        try_run(context, names)
    File "/home/ichimonji10/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
        return func()
    File "/home/ichimonji10/code/robottelo/tests/foreman/api/test_contentview.py", line 864, in setUpClass
        releasever='7Server',
    File "robottelo/api/utils.py", line 118, in enable_rhrepo_and_fetchid
        reposet_id = entities.Product(id=prd_id).fetch_reposet_id(name=reposet)
    File "/home/ichimonji10/code/nailgun/nailgun/entities.py", line 2620, in fetch_reposet_id
        "The length of the results is:", len(results))
    APIResponseError: ('The length of the results is:', 0)

    ======================================================================
    ERROR: @Test: Sync RedHat Repository.
    ----------------------------------------------------------------------
    Traceback (most recent call last):
    File "/home/ichimonji10/code/robottelo/tests/foreman/api/test_repository.py", line 257, in test_redhat_sync_1
        '7Server',
    File "robottelo/api/utils.py", line 118, in enable_rhrepo_and_fetchid
        reposet_id = entities.Product(id=prd_id).fetch_reposet_id(name=reposet)
    File "/home/ichimonji10/code/nailgun/nailgun/entities.py", line 2620, in fetch_reposet_id
        "The length of the results is:", len(results))
    APIResponseError: ('The length of the results is:', 0)

    ----------------------------------------------------------------------
    Ran 841 tests in 4704.321s

    FAILED (SKIP=236, errors=2)

### CODE REVIEW

**Comment by Ichimonji10** [comment]
This PR makes it *much* easier for NailGun to deal with the various kinds of responses that the server returns. I've added a little bit of logic in to the `EntityReadMixin`, and that extra little bit of logic lets me remove boatloads of redundant code from `nailgun/entities.py`. This is apparent in the fact that this PR shrinks NailGun's total size - `+243, -383`!

File: setup.py [hunk 1]
```diff
@@ -27,5 +27,11 @@
         'Programming Language :: Python :: 3.4',
     ],
     packages=find_packages(),
-    install_requires=['fauxfactory', 'pyxdg', 'requests>=2.7', 'setuptools'],
+    install_requires=[
+        'fauxfactory',
+        'inflector',
```
Comment (Ichimonji10) [inline setup.py]: This is a new dependency. The "inflector" library contains a port of the pluralization engine that Ruby on Rails uses. The use of this pluralization engine makes NailGun *much* more capable when dealing with server responses.

File: nailgun/entity_mixins.py [hunk 2]
```diff
@@ -204,6 +205,77 @@ def _get_server_config():
     return config.ServerConfig.get()
 
 
+def _get_entity_id(field_name, attrs):
```
Comment (Ichimonji10) [inline nailgun/entity_mixins.py]: This method's docstring is worth reading. It provides examples of the different ways in which the server might hand back a response. This method is at the heart of this PR.

File: nailgun/entity_mixins.py [hunk 3]
```diff
@@ -204,6 +205,77 @@ def _get_server_config():
...
+        return attrs[field_name_id]
+    else:
+        raise MissingValueError(
+            'Cannot find a value for the "{}" field. Searched for keys named '
+            '{}, but available keys are {}.'
+            .format(field_name, (field_name, field_name_id), attrs.keys())
+        )
+
+
+def _get_entity_ids(field_name, attrs):
```
Comment (Ichimonji10) [inline nailgun/entity_mixins.py]: This method's docstring is worth reading. It provides examples of the different ways in which the server might hand back a response. This method is at the heart of this PR.

File: setup.py (hunk 1 above)
Comment (elyezer) [inline setup.py]: :+1:

**Comment by elyezer** [comment]
APT, that extra logic really saved much duplicated code. Thank you.

**Comment by Ichimonji10** [comment]
Glad you like it! This'll save us quite a bit of frustration in the future.

**Comment by elyezer** [comment]
Indeed.

File: nailgun/entity_mixins.py [hunk 4]
```diff
@@ -204,6 +205,77 @@ def _get_server_config():
...
+    if field_name in attrs:
+        if attrs[field_name] is None:
+            return None
+        else:
+            return attrs[field_name]['id']
+    elif field_name_id in attrs:
+        return attrs[field_name_id]
+    else:
+        raise MissingValueError(
+            'Cannot find a value for the "{}" field. Searched for keys named '
```
Comment (omaciel) [inline nailgun/entity_mixins.py]: Don't we need positional parameters for python 2.6? Or maybe the question should be "don't we want to keep support for python 2.6?

**Comment by omaciel** [comment]
ACK pending comment

File: nailgun/entity_mixins.py (hunk 4 above)
Comment (Ichimonji10) [inline nailgun/entity_mixins.py]: I've never actually made any effort to support Python 2.6. The [`.travis.yml`](https://github.com/SatelliteQE/nailgun/blob/master/.travis.yml#L2-L5) file currently lists Python 2.7, 3.3 and 3.4 as targets. If you'd like, I can try making a new branch and seeing what sorts of changes would be necessary to support 2.6.

**Comment by omaciel** [comment]
Safe to merge?

**Comment by Ichimonji10** [comment]
The coverage decrease is just due to me removing code.

**Comment by omaciel** [comment]
ACK

**Comment by omaciel** [comment]
Gahhh can't merge from phone :/

File: nailgun/entity_mixins.py (hunk 4 above)
Comment (elyezer) [inline nailgun/entity_mixins.py]: I think that it will not worth supporting python 2.6

**Pull request merged**